OutOfSchool/OutOfSchool.WebApi/Services/Gcp/GcpImagesStorageSynchronizationService.cs

```diff
@@ -0,0 +1,138 @@
+            this.imagesStorage = imagesStorage ?? throw new ArgumentNullException(nameof(imagesStorage));
+            this.gcpImagesSyncDataRepository = gcpImagesSyncDataRepository ?? throw new ArgumentNullException(nameof(gcpImagesSyncDataRepository));
+
+            GetAllSyncFunctions = new List<Func<IEnumerable<string>, Task<List<string>>>>
+            {
+                // Entity images
+                gcpImagesSyncDataRepository.GetIntersectWorkshopImagesIds,
+                gcpImagesSyncDataRepository.GetIntersectProviderImagesIds,
+
+                // Entity cover images
+                gcpImagesSyncDataRepository.GetIntersectWorkshopCoverImagesIds,
+                gcpImagesSyncDataRepository.GetIntersectTeacherCoverImagesIds,
+                gcpImagesSyncDataRepository.GetIntersectProviderCoverImagesIds,
+            };
+        }
+
+        private List<Func<IEnumerable<string>, Task<List<string>>>> GetAllSyncFunctions { get; }
+
+        /// <inheritdoc/>
+        public async Task SynchronizeAsync(CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                logger.LogDebug("Gcp storage synchronization was started");
+
+                // gcp returns data by local time
+                var dateTime = DateTime.Now.AddMinutes(-1);
+
+                var listsOfObjects = GetListsOfObjects().ConfigureAwait(false);
+
+                await foreach (var objects in listsOfObjects.WithCancellation(cancellationToken))
+                {
+                    if (objects.Items is null || objects.Items.Count == 0)
+                    {
+                        break;
+                    }
+
+                    var mappedObjects = objects.Items
+                        .Where(x => x.TimeCreated < dateTime)
+                        .Select(x => x.Name)
+                        .ToHashSet();
+
+                    var notAttachedIds = await SynchronizeAll(mappedObjects).ConfigureAwait(false);
+                    await RemoveNotAttachedIds(notAttachedIds).ConfigureAwait(false);
+                }
+            }
+            catch (Exception ex)
+            {
+                logger.LogCritical(ex, "Gcp storage synchronization was interrupted by error");
+            }
+            finally
+            {
+                logger.LogDebug("Gcp storage synchronization was finished");
+            }
+        }
+
+        private async Task<HashSet<string>> SynchronizeAll(HashSet<string> searchIds)
+        {
+            foreach (var syncFunction in GetAllSyncFunctions)
+            {
+                var syncResult = await syncFunction(searchIds).ConfigureAwait(false);
+                if (syncResult != null)
+                {
+                    searchIds.ExceptWith(syncResult);
+                }
+            }
+
+            return searchIds;
+        }
+
+        private IAsyncEnumerable<Objects> GetListsOfObjects()
+        {
+            var options = new ListObjectsOptions
```

Comment:
not a call to immediate action. maybe we can store different "image types" (workshop cover, workshop images, teacher avatars, etc.) under different prefixes and? or go even deeper and prefix workshops for each ministry, etc. It will probably add some complexity to create/update, but will considerably reduce the amount of data that we get during cleanup job. And we can have different jobs for each prefix